From e25a15ed2ec5d9e8c556b7209b71ce7e13aad142 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:02:14 -0400
Subject: [fs] btrfs: wake up the tasks that wait for the io earlier

Message-id: <1407179109-13714-248-git-send-email-sandeen@redhat.com>
Patchwork-id: 87308
O-Subject: [RHEL7.1 PATCH 247/422] Btrfs: wake up the tasks that wait for the io earlier
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Miao Xie <miaox@cn.fujitsu.com>

commit af7a65097b3f0a63caf1755df78d04e1a33588ef
Author: Miao Xie <miaox@cn.fujitsu.com>
Date:   Thu Mar 6 13:54:56 2014 +0800

    Btrfs: wake up the tasks that wait for the io earlier

    The tasks that wait for the IO_DONE flag just care about the io of the dirty
    pages, so it is better to wake up them immediately after all the pages are
    written, not the whole process of the io completes.

    Signed-off-by: Miao Xie <miaox@cn.fujitsu.com>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/ordered-data.c b/fs/btrfs/ordered-data.c
index 73de19c..a75eaa2 100644
--- a/fs/btrfs/ordered-data.c
+++ b/fs/btrfs/ordered-data.c
@@ -349,10 +349,13 @@ int btrfs_dec_test_first_ordered_pending(struct inode *inode,
  if (!uptodate)
   set_bit(BTRFS_ORDERED_IOERR, &entry->flags);
 
- if (entry->bytes_left == 0)
+ if (entry->bytes_left == 0) {
   ret = test_and_set_bit(BTRFS_ORDERED_IO_DONE, &entry->flags);
- else
+  if (waitqueue_active(&entry->wait))
+   wake_up(&entry->wait);
+ } else {
   ret = 1;
+ }
 out:
  if (!ret && cached && entry) {
   *cached = entry;
@@ -410,10 +413,13 @@ have_entry:
  if (!uptodate)
   set_bit(BTRFS_ORDERED_IOERR, &entry->flags);
 
- if (entry->bytes_left == 0)
+ if (entry->bytes_left == 0) {
   ret = test_and_set_bit(BTRFS_ORDERED_IO_DONE, &entry->flags);
- else
+  if (waitqueue_active(&entry->wait))
+   wake_up(&entry->wait);
+ } else {
   ret = 1;
+ }
 out:
  if (!ret && cached && entry) {
   *cached = entry;
-- 
1.7.1